File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays.DotNet/director/Chunks/RaysKeyTableChunk.cs
﻿using Microsoft.Extensions.Logging;
using ProjectorRays.Common;
using ProjectorRays.Director;

namespace ProjectorRays.director.Chunks;

public class RaysKeyTableChunk : RaysChunk
{
    public List<KeyTableEntry> Entries = new();

    public RaysKeyTableChunk(RaysDirectorFile? dir) : base(dir, ChunkType.KeyTableChunk) { }

    public override void Read(ReadStream stream)
    {
        // Use the movie's endianness to decode the key table correctly.
        stream.Endianness = Dir?.Endianness ?? Endianness.BigEndian;
        ushort entrySize = stream.ReadUint16();
        stream.Skip(2);
        uint count = stream.ReadUint32();
        uint used = stream.ReadUint32();
        for (int i = 0; i < used; i++)
        {
            var entry = new KeyTableEntry();
            entry.Read(stream);
            Entries.Add(entry);
        }
        Dir?.Logger.LogInformation($"KeyTableChunk: EntrySize={entrySize}, Count={count}, Used={used}, ParsedEntries={Entries.Count}");
        //Dir?.Logger.LogInformation("KeyTableChunk: Entries:"+string.Join('|', Entries.Select(e => $"{e.LogString()}")));
    }

    public override void WriteJSON(RaysJSONWriter json)
    {
        json.StartObject();
        json.WriteField("entries", Entries.Count);
        json.WriteKey("entriesList");
        json.StartArray();
        foreach (var e in Entries)
        {
            e.WriteJSON(json);
        }
        json.EndArray();
        json.EndObject();
    }
}

File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays_Python/director/chunks/rays_key_table_chunk.py
from .rays_list_chunk import RaysListChunk
from .rays_chunk import ChunkType
from ...common.stream import ReadStream
from ..rays_subchunk import KeyTableEntry

class RaysKeyTableChunk(RaysListChunk):
    def __init__(self, dir=None):
        super().__init__(dir, ChunkType.KeyTableChunk)
        self.entries = []

    def read(self, stream: ReadStream):
        stream.set_endianness(self.dir.endianness if self.dir else stream.endianness)
        self.entry_size = stream.read_uint16()
        stream.skip(2)
        self.count = stream.read_uint32()
        self.used = stream.read_uint32()
        for _ in range(self.used):
            entry = KeyTableEntry()
            entry.read(stream)
            self.entries.append(entry)
        if self.dir:
            self.dir._logger.info(
                "KeyTableChunk: EntrySize=%d, Count=%d, Used=%d, ParsedEntries=%d",
                self.entry_size,
                self.count,
                self.used,
                len(self.entries),
            )

    def write_json(self, writer):
        writer.start_object()
        writer.write_key('count'); writer.write_val(len(self.entries))
        writer.end_object()
